### PR TITLE
 HLE/Services: Allow specifying a SessionData template parameter to ServiceFramework.

### PR DESCRIPTION
**Pull request by Subv**
Some services can have multiple clients at the same time, and they identify the different clients using the server session as a key.
This parameter (if present) should be a structure that contains the per-session data for each service.
The data can be retrieved using ServiceFramework::GetSessionData(session)

cc @wwylele

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3284)

### CODE REVIEW

File: src/core/hle/kernel/hle_ipc.h [hunk 1]
```diff
@@ -56,11 +57,34 @@ class SessionRequestHandler : public std::enable_shared_from_this<SessionRequest
...
-    /// A ServerSession whose server endpoint is an HLE implementation is kept alive by this list
-    // for the duration of the connection.
-    std::vector<SharedPtr<ServerSession>> connected_sessions;
+    /// Creates the storage for the session data of the service.
+    virtual std::unique_ptr<SessionDataBase> MakeSessionData() const = 0;
+
+    /// Returns the session data associated with the server session.
+    template <typename T>
+    T* GetSessionData(SharedPtr<ServerSession> session) {
+        static_assert(std::is_base_of<SessionDataBase, T>::value,
```
Comment (lioncash) [inline src/core/hle/kernel/hle_ipc.h]:
```cpp
std::is_base_of<SessionDatabase, T>()
```

File: src/core/hle/kernel/hle_ipc.cpp [hunk 2]
```diff
@@ -2,26 +2,34 @@
...
+void SessionRequestHandler::ClientConnected(
+    Kernel::SharedPtr<Kernel::ServerSession> server_session) {
     server_session->SetHleHandler(shared_from_this());
-    connected_sessions.push_back(server_session);
+    connected_sessions.emplace_back(std::move(server_session), MakeSessionData());
 }
 
-void SessionRequestHandler::ClientDisconnected(SharedPtr<ServerSession> server_session) {
+void SessionRequestHandler::ClientDisconnected(
+    Kernel::SharedPtr<Kernel::ServerSession> server_session) {
```
Comment (wwylele) [inline src/core/hle/kernel/hle_ipc.cpp]: No need for `Kernel::`

File: src/core/hle/kernel/hle_ipc.cpp [hunk 3]
```diff
@@ -2,26 +2,34 @@
...
 
 namespace Kernel {
 
-void SessionRequestHandler::ClientConnected(SharedPtr<ServerSession> server_session) {
+SessionRequestHandler::SessionInfo::SessionInfo(SharedPtr<ServerSession> session,
+                                                std::unique_ptr<SessionDataBase> data)
+    : session(std::move(session)), data(std::move(data)) {}
+
+void SessionRequestHandler::ClientConnected(
+    Kernel::SharedPtr<Kernel::ServerSession> server_session) {
```
Comment (wwylele) [inline src/core/hle/kernel/hle_ipc.cpp]: No need for `Kernel::`

File: src/core/hle/service/fs/archive.h [hunk 4]
```diff
@@ -108,6 +101,10 @@ class Directory final : public Kernel::SessionRequestHandler {
 
 protected:
     void HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session) override;
+
+    std::unique_ptr<SessionDataBase> MakeSessionData() const override {
+        return nullptr;
```
Comment (wwylele) [inline src/core/hle/service/fs/archive.h]: Shouldn't this return a pointer to a new `FileSessionSlot` isntance?

File: src/core/hle/kernel/hle_ipc.h [hunk 5]
```diff
@@ -47,7 +48,7 @@ class SessionRequestHandler : public std::enable_shared_from_this<SessionRequest
      * associated ServerSession alive for the duration of the connection.
      * @param server_session Owning pointer to the ServerSession associated with the connection.
      */
-    void ClientConnected(SharedPtr<ServerSession> server_session);
+    virtual void ClientConnected(SharedPtr<ServerSession> server_session);
```
Comment (wwylele) [inline src/core/hle/kernel/hle_ipc.h]: What's the reason to make this virtual?

File: src/core/hle/service/fs/archive.h (hunk 4 above)
Comment (Subv) [reply src/core/hle/service/fs/archive.h]: This is the Directory class

Comment (wwylele) [reply src/core/hle/service/fs/archive.h]: oops, but in this case why is this needed?

**Pull request merged**